apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfPublisher.cs

```diff
@@ -0,0 +1,626 @@
+            {
+                if (_shutdownLock.Locked(() => {
+                    var shutdownStarted = _shutdownStarted;
+                    _shutdownStarted = true;
+                    return !shutdownStarted;
+                }))
+                {
+                    // All batches sent and shutdown requested, so signal shutdown completed successfully.
+                    _taskHelper.Run(async () =>
+                    {
+                        if (_shutdown != null)
+                        {
+                            await _taskHelper.ConfigureAwaitHideErrors(_shutdown());
+                        }
+                        if (_hardStopCts.IsCancellationRequested)
+                        {
+                            _shutdownTcs.SetCanceled();
+                        }
+                        else
+                        {
+                            _shutdownTcs.SetResult(0);
+                        }
+                    });
+                }
+            }
+        }
+
+        // Not "async void" to gain standard async exeption handling
+        private async Task CreateBatch()
+        {
+            // Pre-condition: Must be locked - lock active up until first 'await'.
+            _currentBatch = new Batch();
+            var timerToken = _currentBatch.TimerCts.Token;
+            // After time-delay threshold, send the batch regardless of fullness.
+            await _taskHelper.ConfigureAwait(_scheduler.Delay(_batchDelayThreshold, timerToken));
+            // If batch has already moved to queue, timerToken will have been cancelled.
+            lock (_lock)
+            {
+                // Check for cancellation inside lock to avoid race-condition.
+                if (!timerToken.IsCancellationRequested)
+                {
+                    // Force queuing of the current batch, whatever the size.
+                    // There will always be at least one message in the batch.
+                    QueueCurrentBatchIfRequired(true, 0);
+                }
+            }
+        }
+
+        private void QueueCurrentBatchIfRequired(bool force, int extraByteCount)
+        {
+            // Pre-condition: Must be locked
+            if (_currentBatch != null && (force || CurrentBatchIsFull(_idleClients.Count == 0, extraByteCount)))
+            {
+                // Cancel the timeout for this batch.
+                _currentBatch.TimerCts.Cancel();
+                // Queue the batch ready for sending.
+                _batchesReady.Enqueue(_currentBatch);
+                // Mark that there is no current batch.
+                _currentBatch = null;
+                // Trigger send to server.
+                Task unusedSendTask = TriggerSend();
+            }
+        }
+
+        // Pre-condition: Must be locked
+        // Allow one message that is larger than requestByteThreshold
+        private bool CurrentBatchIsFull(bool isQueuing, int extraByteCount) => isQueuing ?
```

Comment:
I'm finding this very hard to understand. Rather than struggle through it, could you elaborate in a comment, or direct the reader to an existing explanation?